### PR TITLE
Rewrote plugin. See README.md for more info.

### PR DESCRIPTION
**Pull request by QWp6t**
* Update to Apache Server Configs v2.14.0
* Rewrote plugin as a class
* Rewrote logic so that server and WordPress configurations are only checked when plugin is actively in use
* Added filters to provide users with more control over the rules that are being applied
* Added a dedicated WordPress section to the bottom of h5bp-htaccess

### CODE REVIEW

**Comment by QWp6t** [comment]
This closes the other two issues, by the way.

File: wp-h5bp-htaccess.php [hunk 1]
```diff
@@ -1,46 +1,113 @@
...
-/**
- * Add HTML5 Boilerplate's .htaccess via WordPress
- */
-function roots_add_h5bp_htaccess($content) {
-  global $wp_rewrite;
-  $home_path = function_exists('get_home_path') ? get_home_path() : ABSPATH;
-  $htaccess_file = $home_path . '.htaccess';
-  $mod_rewrite_enabled = function_exists('got_mod_rewrite') ? got_mod_rewrite() : false;
+  private static $instance;
+  private static $marker = 'Roots Server Config';
```
Comment (swalkinshaw) [inline wp-h5bp-htaccess.php]: Think a `const` is more appropriate here?

```php
const MARKER = 'Roots Server Config';
```

File: wp-h5bp-htaccess.php [hunk 2]
```diff
@@ -1,46 +1,113 @@
...
+  }
+  
+  protected function __construct() {
+    require_once ABSPATH . '/wp-admin/includes/file.php';
+    require_once ABSPATH . '/wp-admin/includes/misc.php';
+    
+    self::$mod_rewrite_enabled = got_mod_rewrite();
+    self::$home_path = get_home_path();
+    self::$wp_htaccess_file = self::$home_path . '.htaccess';
+    self::$roots_htaccess_file = locate_template(['server_configs.conf','h5bp-htaccess.conf']);
```
Comment (swalkinshaw) [inline wp-h5bp-htaccess.php]: space after comma

File: wp-h5bp-htaccess.php [hunk 3]
```diff
@@ -1,46 +1,113 @@
...
+      'www.example.com/' => $host . $path,
+      'www.example.com'  => $host,
+      '/wordpress' => parse_url(site_url(), PHP_URL_PATH)
+    ];
+    self::$rules_filters = apply_filters('roots/h5bp-htaccess-filters', self::$rules_filters);
+  }
+  
+  protected function verifySetup() {
+    if (!get_option('permalink_structure')) {
+      $this->errors[] = sprintf(__('Please enable %s.', 'roots'), '<a href="' . admin_url('options-permalink.php') . '">Permalinks</a>');
```
Comment (swalkinshaw) [inline wp-h5bp-htaccess.php]: Can you make `'root'` a class const to DRY it up?

File: wp-h5bp-htaccess.php [hunk 4]
```diff
@@ -1,46 +1,113 @@
...
+    
+    if (!file_exists(self::$roots_htaccess_file)) {
+      $this->errors[] = sprintf(__('Cannot find the file %s.', 'roots'), self::$roots_htaccess_file);
+    }
+    
+    return empty($this->errors);
+  }
+  
+  protected function isWritable() {
+    return is_writable(self::$home_path) || file_exists(self::$wp_htaccess_file) && is_writable(self::$wp_htaccess_file);
```
Comment (swalkinshaw) [inline wp-h5bp-htaccess.php]: This could use some extra parantheses for grouping to make the intention clearer.

File: wp-h5bp-htaccess.php [hunk 5]
```diff
@@ -1,46 +1,113 @@
...
-  if (!is_writable(get_home_path() . '.htaccess')) {
-    if (current_user_can('administrator')) {
-      add_action('admin_notices', create_function('', "echo '<div class=\"error\"><p>" . sprintf(__('Please make sure your <a href="%s">.htaccess</a> file is writable ', 'roots'), admin_url('options-permalink.php')) . "</p></div>';"));
-    }
-  }
-}
-add_action('admin_init', 'roots_htaccess_writable');
+add_action('generate_rewrite_rules', __NAMESPACE__ . '\\ApacheServerConfig::init', 0, -9999);
+
+class ApacheServerConfig {
```
Comment (swalkinshaw) [inline wp-h5bp-htaccess.php]: kill this blank line

File: wp-h5bp-htaccess.php [hunk 6]
```diff
@@ -1,46 +1,113 @@
...
+  }
+  
+  public function alerts() {
+    $alert = function($message) {
+      echo '<div class="error"><p>' . $message . '</p></div>';
+    };
+    if (current_user_can('administrator')) {
+      add_action('admin_notices', function() use ($alert) {
+        array_map($alert, $this->errors);
+      });
```
Comment (swalkinshaw) [inline wp-h5bp-htaccess.php]: I like how the errors are handled :+1:

**Comment by QWp6t** [comment]
Done.

**Comment by swalkinshaw** [comment]
:+1: assuming it works!

**Comment by QWp6t** [comment]
I've tested in a Bedrock environment and I just tested it in a traditional WordPress environment. Worked for me on both of those. I also tested all of the examples I used in README.md.

Everything checks out on my end.

**Pull request merged**